src/modules/rlm_ldap/user.c

```diff
@@ -35,6 +35,136 @@ USES_APPLE_DEPRECATED_API
 
 #include "rlm_ldap.h"
 
+/** Holds state of user searches in progress
+ *
+ */
+typedef struct {
+	rlm_ldap_t const	*inst;
+	fr_ldap_thread_trunk_t	*ttrunk;
+	char const		*base_dn;
+	char const		*filter;
+	char const * const	*attrs;
+	fr_ldap_query_t		*query;
+	fr_ldap_query_t		**out;
+} ldap_user_find_ctx;
+
+static unlang_action_t ldap_find_user_async_result(rlm_rcode_t *p_result, UNUSED int *priority, request_t *request, void *uctx)
+{
+	ldap_user_find_ctx	*user_ctx = talloc_get_type_abort(uctx, ldap_user_find_ctx);
+	fr_ldap_query_t		*query = user_ctx->query;
+	LDAPMessage		*entry;
+	int			cnt, ldap_errno;
+	char			*dn;
+	fr_pair_t		*vp;
+
+	if (!user_ctx->inst->userobj_sort_ctrl) {
+		cnt = ldap_count_entries(query->ldap_conn->handle, query->result);
+		if (cnt > 1) {
+			REDEBUG("Ambiguous search result, returned %i unsorted entries (should return 1 or 0).  "
+				"Enable sorting, or specify a more restrictive base_dn, filter or scope", cnt);
+			REDEBUG("The following entries were returned:");
+			RINDENT();
+			for (entry = ldap_first_entry(query->ldap_conn->handle, query->result);
+			     entry;
+			     entry = ldap_next_entry(query->ldap_conn->handle, entry)) {
+				dn = ldap_get_dn(query->ldap_conn->handle, entry);
+				REDEBUG("%s", dn);
+				ldap_memfree(dn);
+			}
+			REXDENT();
+			RETURN_MODULE_INVALID;
+		}
+	}
+
+	entry = ldap_first_entry(query->ldap_conn->handle, query->result);
+	if (!entry) {
+		ldap_get_option(query->ldap_conn->handle, LDAP_OPT_RESULT_CODE, &ldap_errno);
+		REDEBUG("Failed retrieving entry: %s",
+			ldap_err2string(ldap_errno));
+
+		RETURN_MODULE_FAIL;
+	}
+
+	dn = ldap_get_dn(query->ldap_conn->handle, entry);
+	if (!dn) {
+		ldap_get_option(query->ldap_conn->handle, LDAP_OPT_RESULT_CODE, &ldap_errno);
+		REDEBUG("Retrieving object DN from entry failed: %s", ldap_err2string(ldap_errno));
+
+		RETURN_MODULE_FAIL;
+	}
+	fr_ldap_util_normalise_dn(dn, dn);
+
+	RDEBUG2("User object found at DN \"%s\"", dn);
+
+	MEM(pair_update_control(&vp, attr_ldap_userdn) >= 0);
+	fr_pair_value_strdup(vp, dn, false);
+	ldap_memfree(dn);
+	if (user_ctx->out) *user_ctx->out = user_ctx->query;
+
+	return UNLANG_ACTION_CALCULATE_RESULT;
```

Comment:
I think/hope the interpreter records the last rcode in `request->rcode` so we can get at it easily, otherwise there's going to need to be some way of explicitly passing a result back in the uctx.